### PR TITLE
u32-backed OneIndexed

### PR DESCRIPTION
**Pull request by youknowone**
I tried to minimize diagnostic side code changes.

The key change of interface is `to_one_indexed`. By adding this method, the interface consistently requires to use `one_indexed` or `zero_indexed` to convert `OneIndexed` to usize or vice versa.

### CODE REVIEW

File: crates/ruff_python_ast/src/source_code/line_index.rs [hunk 1]
```diff
@@ -236,49 +236,71 @@ impl IndexKind {
...
 impl OneIndexed {
     // SAFETY: These constants are being initialized with non-zero values
     /// The smallest value that can be represented by this integer type.
     pub const MIN: Self = unwrap(Self::new(1));
     /// The largest value that can be represented by this integer type
-    pub const MAX: Self = unwrap(Self::new(usize::MAX));
+    pub const MAX: Self = unwrap(Self::new(u32::MAX));
 
-    pub const ONE: NonZeroUsize = unwrap(NonZeroUsize::new(1));
+    const ONE: NonZeroU32 = unwrap(NonZeroU32::new(1));
```
Comment (youknowone) [inline crates/ruff_python_ast/src/source_code/line_index.rs:265]: This is unrelated to the described change. I suggest to remove `pub` here or rename `ONE` to something other like `NONZERO_ONE`. The name `ONE` makes it looks like `Self` type, but it isn't.

**Review by MichaReiser — COMMENTED** [review]
Thanks for this contribution. 

I am conflicted about whether this is worth changing. My first draft of `OneIndex` used `NonZeroU32` internally but I then decided to change it to `NonZeroUsize` because:

* It is easier to use (no panics, `unwrap`, casts or conversions)
* Other than `TextSize`, `OneIndexed` isn't used in any performant critical path where optimizing its size matters. The only place that comes to my mind that may improve from this change is when using the `JsonEmitter` with a large number of `Diagnostic`s, but I would reason that the lookup of the source location is the dominating performance factor. 
* Assertions (potential panics) have a small runtime cost. 

What was your motivation for this change? Is this infrastructure that you plan to ship with RustPython and you want to make sure they're aligned?

File: crates/ruff_python_ast/src/source_code/line_index.rs [hunk 2]
```diff
@@ -236,49 +236,71 @@ impl IndexKind {
...
+pub struct OneIndexed(NonZeroU32);
+
+#[allow(clippy::cast_possible_truncation)] // manually checked
+const fn index_to_u32(value: usize) -> u32 {
+    assert!(
+        value <= u32::MAX as usize,
+        "Value is too large to be represented as a OneIndexed"
+    );
+    value as u32
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/ruff_python_ast/src/source_code/line_index.rs:256]: I got the second reason, but have a question about the first reason.

What happens if source code is bigger than 4GB? Because TextSize is backed by u32, I thought source code bigger than 4GB will not be located.
Does ruff use different mechanism to solve large files?
Whether it is or not, the exact 4GB file seem to be handled in same way as 4GB+ files.

----

Not strictly related to this topic, but some platforms even take 64bit file size when the underlying architecture is 32bit. Source code length can be not limited to u32. (Though I hope nobody tries to make 4GB source code in single file)

File: crates/ruff_python_ast/src/source_code/line_index.rs [hunk 3]
```diff
@@ -236,49 +236,71 @@ impl IndexKind {
...
             Some(value) => Some(Self(value)),
             None => None,
         }
     }
 
     /// Construct a new [`OneIndexed`] from a zero-indexed value
+    ///
+    /// # Panics
+    /// Panics if the given value doesn't fit in u32
     pub const fn from_zero_indexed(value: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/ruff_python_ast/src/source_code/line_index.rs:279]: I agree making this function to take u32 is better.

About `to_one_indexed`, I will follow your decision, but I'd like to add one more comment.

Unlike `NonZero` type, which means zero is out of range, `OneIndexed` offers conversion functions to zero indexed value. I also can say `OneIndexed` is another representation of zero indexed value.
When user convert this value to usize, sometimes it is conversion to underlying value (zero-indexed) and something it is conversion to representation (one-indexed).
By providing two symmetric interfaces, users can always select their intentions with equal weight by looking at the function names. It helps to reduce mistakes.
By providing `From<OneIndexed> for usize`, users tends to regards one way is default behavior and the other way is optional function.  It may be the truth because this type is designed to hold one-indexed values. But optional function is easy to be forgotten.
Practically it might be confusing. It will not be confused in one way - using to_zero_indexed is very explicit, but the other way - writing `value.into()` for `usize` conversion - is very easy to happen inconsiously.

Ruff codebase is using either conversion in similar ratio in small numbers - to-zero 8 and to-one 7. Implementing `From<OneIndexed> for usize` doesn't give much value and also hard to say which conversion is dominant. I personally will not provide that `From` implementation while it supports both one and zero indexed conversions.

File: crates/ruff_python_ast/src/source_code/line_index.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_ast/src/source_code/line_index.rs:256]: > What happens if source code is bigger than 4GB? Because TextSize is backed by u32, I thought source code bigger than 4GB will not be located.

Ruff will panic for files larger than 4GB. We should probably add a nice diagnostic right before parsing the file to gracefully handle this case.

The main point here is that `TextSize` can handle files up to `u32::MAX`, but `OneIndexed` only up to `u32::MAX - 1`.

File: crates/ruff_python_ast/src/source_code/line_index.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_ast/src/source_code/line_index.rs:279]: > About to_one_indexed, I will follow your decision, but I'd like to add one more comment.

That's an excellent point. I guess what I find confusing is that both `get` and `to_one_indexed` return the same value and only differ by the return type. What do you think of `get` and `get_usize()` (or `to_one_indexed()` and `to_one_indexed_usize()`?

File: crates/ruff_python_ast/src/source_code/line_index.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/ruff_python_ast/src/source_code/line_index.rs:256]: That's exactly what I meant to. If ruff reject 4GB+ files, ruff can reject the exact 4GB file too when reading file if length  = u32::MAX. If ruff is going to add a nice diagnostic, it can add the same thing by checking length == u32::MAX when reading file length.
It will simplify exception handling just by excluding one very rare edge case.

File: crates/ruff_python_ast/src/source_code/line_index.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (youknowone) [reply crates/ruff_python_ast/src/source_code/line_index.rs:279]: Otherwise, just full matrix also makes sense to me.

`to_one_indexed() -> u32`
`to_zero_indexed() -> u32`
`to_one_indexed_usize() -> usize`
`to_zero_indexed_usize() -> usize`

[the remark this answers is not in the document]
Comment (youknowone) [reply crates/ruff_python_ast/src/source_code/line_index.rs:279]: Otherwise, just full matrix also makes sense to me.

`to_one_indexed() -> u32`
`to_zero_indexed() -> u32`
`to_one_indexed_usize() -> usize`
`to_zero_indexed_usize() -> usize`

File: crates/ruff/src/message/diff.rs [hunk 4]
```diff
@@ -55,8 +55,9 @@ impl Display for Diff<'_> {
             .map(|op| (op.old_range().start, op.new_range().start))
             .unwrap_or_default();
 
-        let digit_with =
-            calculate_print_width(OneIndexed::from_zero_indexed(largest_new.max(largest_old)));
+        let digit_with = calculate_print_width(
+            OneIndexed::try_from_zero_indexed(largest_new.max(largest_old)).unwrap(),
```
Comment (youknowone) [inline crates/ruff/src/message/diff.rs:59]: I left error case as unwrap. You will be able to fill proper handling later.

File: crates/ruff_python_ast/src/source_code/line_index.rs [hunk 5]
```diff
@@ -236,49 +246,76 @@ impl IndexKind {
...
+    }
+
     /// Return the zero-indexed primitive value for this [`OneIndexed`]
-    pub const fn to_zero_indexed(self) -> usize {
+    pub const fn to_zero_indexed(self) -> u32 {
         self.0.get() - 1
     }
 
+    /// Return the zero-indexed usize value for this [`OneIndexed`]
+    pub const fn to_zero_indexed_usize(self) -> usize {
```
Comment (youknowone) [inline crates/ruff_python_ast/src/source_code/line_index.rs:310]: Added a new method not to use `as usize` everywhere.

**Review by MichaReiser — CHANGES_REQUESTED** [review]
Could you explain your main motivation behind my change? 

I remain sceptical if performance is the main reason because it increases the complexity of the API, adds more paths that can potentially panic without it measurably increasing performance (because we don't use `SourceLocation` in a hot path).

**Review by youknowone — COMMENTED** [review]
I need to work on this changes and wanted to contribute it back because it seems worth (even though not that much).
It doesn't need to be aligned. RustPython borrowed a copy of code but it is never used in Ruff side.

For panic points view,
You said `OneIndexed` can be used for non-text-size context, but it was always used for text size directly or indirectly as I've checked `unwrap()` parts.
The main source of unwrap came from TextDiff. I checked it is external dependency but the value will be always under TextSize range because it takes source code limited by TextSize (well, not that confident tbh). Well, the right way to handling this will be separating OneIndexedU32 and OneIndexed(usize). You may not like it because it is unnecessary code expansion for non-critical path.

I agree the change doesn't make sufficient advantage to the project. Please feel free to close it if you think it doesn't fit in the project.

**Comment by MichaReiser** [comment]
I'll close this for now because it increases the complexity of the API without improving performance, memory consumption significantly. 

We should reconsider this change if we start storing many `SourceLocation` nodes or if aligning with `RustPython`'s `SourceLocation` implementation removes the need for Ruff to have it's own fork or eases pulling in upstream changes.
